gnomad_hail/resources/resource_utils.py

```diff
@@ -9,114 +9,221 @@
+        else:
+            return hl.read_matrix_table(self.path)
+
+    def import_resource(self, overwrite: bool = False, **kwargs):
+        """
+        Imports the MatrixTable resource using its import_func and writes it in its path.
+
+        :param overwrite: If set, existing file(s) will be overwritten
+        :param kwargs: Any other parameters to be passed to hl.MatrixTable.write
+        :return: Nothing
+        """
+        self.import_func(**self.import_sources).write(self.path, overwrite=overwrite, **kwargs)
 
 
 class PedigreeResource(BaseResource):
     """
     A pedigree resource
+
+    :param path: The Pedigree path (typically ending in .fam or .ped)
+    :param import_sources: Any sources that are required for the import and need to be kept track of and/or passed to the import_func (e.g. .vcf path for an imported VCF)
+    :param import_func: A function used to import the MatrixTable. `import_func` will be passed the `import_sources` dictionary as kwargs.
+    :param quant_pheno: If ``True``, phenotype is interpreted as quantitative.
+    :param delimiter: Field delimiter regex.
+    :param missing: The string used to denote missing values. For case-control, 0, -9, and non-numeric are also treated as missing.
     """
 
-    def __init__(self, path: str, import_sources: Optional[Dict[str, Any]] = None, ):
+    def __init__(
+            self,
+            path: Optional[str] = None,
+            import_sources: Optional[Dict[str, Any]] = None,
+            import_func: Optional[Callable[..., hl.MatrixTable]] = None,
+            quant_pheno: bool = False,
+            delimiter: str = r"\\s+",
+            missing: str = 'NA'
+    ):
         super().__init__(
             path=path,
             import_sources=import_sources,
-            expected_file_extension='.fam'
+            import_func=import_func,
+            expected_file_extensions=['.fam', '.ped']
         )
 
-    def ht(self, delimiter=r"\\s+") -> hl.Table:
+        self.quant_pheno = quant_pheno
+        self.delimiter = delimiter
+        self.missing = missing
+
+    def ht(self) -> hl.Table:
         """
         Reads the pedigree into a family HT using hl.import_fam().
 
-        :param delimiter: Delimiter used in the ped file
         :return: Family table
         """
-        return hl.import_fam(self.path, delimiter=delimiter)
+        return hl.import_fam(self.path, quant_pheno=self.quant_pheno, delimiter=self.delimiter, missing=self.missing)
 
-    def pedigree(self, delimiter=r"\\s+") -> hl.Pedigree:
+    def pedigree(self) -> hl.Pedigree:
         """
         Reads the pedigree into an hl.Pedigree using hl.Pedigree.read().
 
         :param delimiter: Delimiter used in the ped file
         :return: pedigree
         """
-        return hl.Pedigree.read(self.path, delimiter=delimiter)
+        return hl.Pedigree.read(self.path, delimiter=self.delimiter)
+
+    def import_resource(self):
```

Comment:
Since the `import_resource` method on most resource classes' (Table, MatrixTable, and BlockMatrix) has an overwrite option and passes kwargs to the underlying Hail write function, I would expect this to do the same.

Reply:
`Pedigree.write` takes no arguments. It WILL overwrite (I'll add this as a comment and we _could_ add an `overwrite` arg with default `True` (which would be different from other resources) and throw a `NotImplementedError` if it is set to `False`. We could also add the `kwargs` to future-proof things although atm no argument is supported.

Reply:
Since `import_resource` is a method on the base resource class, it should have the same signature for all resource classes. If you have a collection of resources, you should be able to iterate over them and call `import_resource` without worrying about the details of its implementation in each subclass. For PedigreeResource, I think `overwrite` should default False the same as the other classes and raise (or warn) if it's not set to True.

With the current implementation, if we move most of the code in the two reference data import scripts ([GRCh37](https://github.com/macarthur-lab/gnomad_hail/blob/923fec00302cb06c1b90508b20fe6926c3b25591/gnomad_hail/resources/grch37/reference_data_import.py) and [GRCh38](https://github.com/macarthur-lab/gnomad_hail/blob/923fec00302cb06c1b90508b20fe6926c3b25591/gnomad_hail/resources/grch38/reference_data_import.py)) into the resource definitions using `import_func`, we can't do this:
```python
all_resources = [
   purcell_5k_intervals,
   clinvar,
   na12878_giab,
]

for resource in all_resources:
  resource.import_resource(overwrite=True)
```
because PedigreeResources would have to be special cased.

Reply:
If we add a dumby `overwrite` to pedigree resources couldnt we do what you mention above @nawatts ? Since the pedigree will always overwrite?

Reply:
Yes, we'd need an `overwrite` argument here (doesn't really matter if it's used or not) to avoid a `TypeError: import_resource() got an unexpected keyword argument 'overwrite'` in that example.

Reply:
> Since `import_resource` is a method on the base resource class, it should have the same signature in the base class and in all resource subclasses. If you have a collection of resources, you should be able to iterate over them and call `import_resource` without worrying about the details of its implementation in each subclass. For PedigreeResource, I think `overwrite` should default False the same as the other classes and raise (or warn) if it's not set to True.
> 
> With the current implementation, if we move most of the code in the two reference data import scripts ([GRCh37](https://github.com/macarthur-lab/gnomad_hail/blob/923fec00302cb06c1b90508b20fe6926c3b25591/gnomad_hail/resources/grch37/reference_data_import.py) and [GRCh38](https://github.com/macarthur-lab/gnomad_hail/blob/923fec00302cb06c1b90508b20fe6926c3b25591/gnomad_hail/resources/grch38/reference_data_import.py)) into the resource definitions using `import_func`, we can't do this:
> 
> ```python
> all_resources = [
>    purcell_5k_intervals,
>    clinvar,
>    na12878_giab,
> ]
> 
> for resource in all_resources:
>   resource.import_resource(overwrite=True)
> ```
> 
> because PedigreeResources would have to be special cased.

Haha, I've actually written that already :)

Reply:
I'm fine to add the dummy `overwrite` -- now, I think it needs to be set to `True` by default here

Reply:
For other resources, I feel like keeping the hail default of `overwrite=False` would have made sense but now we're facing a pickle. Maybe no default?

Reply:
With no default set, we'd be using hails default which is `False` right?

Reply:
No, we'd force the user to provide a value

Reply:
I guess a last option, which was suggested by @nawatts, is to only use `kwargs` and pass these to the corresponding hail write function.